lib/verse/sequel/repository.rb

```diff
@@ -107,16 +107,26 @@ def index_impl(
         with_db_mode :r do
           query = filtering.filter_by(scope, filters, self)
 
-          query = query.offset( (page - 1) * items_per_page).limit(items_per_page) if page
-
           sort ||= [
             [self.class.table, self.class.primary_key.to_s].join(".")
           ]
           query = prepare_ordering(query, sort)
 
-          count = query_count ? query.count : nil
+          offset = (page - 1) * items_per_page
+          query = query.offset(offset).limit(items_per_page) if page
+
+          meta_data = {}
+
+          if query_count
+            # For performance: count max 1000 rows from offset
+            count_query = query.limit(1000)
+            limited_count = count_query.count
```

Comment:
This won't work.

Let say you have this query:

```sql
SELECT * FROM employees;
```

You will create this query:

```sql
SELECT COUNT(*) FROM employees OFFSET XXX LIMIT 1000;
```

Instead we want this query:

```sql
SELECT COUNT(*) FROM (SELECT * FROM EMPLOYEES OFFSET XXX LIMIT 1000);
```

Please write some specs, because you haven't tested it; use generate_series to test it.